akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -41,112 +54,166 @@ class AffinityPool(numThreads: Int, affinityGroupSize: Int, tf: ThreadFactory) e
 
   private def tryEnqueue(command: Runnable) = {
     // does a mod on the hash code to determine which queue to put task in
-    val queueIndex = Math.abs(command.hashCode()) & (numThreads - 1)
+    val queueIndex = Math.abs(command.hashCode()) & (parallelism - 1)
     workQueues(queueIndex).add(command)
   }
 
-  private def onWorkerExit(w: ThreadPoolWorker): Unit = {
-    locked(mainLock) {
+  /**
+   * Each worker should go through that method while terminating.
+   * In turn each worker is responsible for modifying the pool
+   * state accordingly. For example if this is the last worker
+   * and the queue is empty and we are in a ShuttingDown state
+   * the worker can transition the pool to ShutDown and attempt
+   * termination
+   *
+   * Furthermore, if this worker has experienced abrupt termination
+   * due to an exception being thrown in user code, the worker is
+   * responsible for adding one more worker to compensate for its
+   * own termination
+   *
+   */
+  private def onWorkerExit(w: ThreadPoolWorker, abruptTermination: Boolean): Unit =
+    locked(bookKeepingLock) {
       workers.remove(w)
+      if (workers.isEmpty && !abruptTermination && poolState >= ShuttingDown) {
+        poolState = ShutDown // transition to shutdown and try to transition to termination
+        attemptPoolTermination()
+      }
+      if (abruptTermination && poolState == Running)
+        addWorker(workers, w.q)
     }
+
+  override def execute(command: Runnable): Unit = {
+    if (command == null)
+      throw new NullPointerException
+    if (!(poolState == Running && tryEnqueue(command)))
```

Comment:
@viktorklang  To be honest I do not think any of that would pose a problem when it comes to shutting down. As described in the comments associated with the states of the service. The moment `poolState` is set to `ShuttingDown` any task that is submitted to the service will be rejected. The workers will gradually drain their queues and one of them will terminate the pool upon its own exit. This is when `shutdown` is called. If `shutdownNow` is called then the pool transitions into `ShutDown` state, no tasks can be enqueued and no workers are polling the queues anymore so whatever was in the queues in terms of tasks is ignored. This is pretty much the same behavior as in `ThreadPoolExecutor`

When it comes to concurrency considerations and we think about race conditions. What can happen is that `T1` calls `shutdown`, acquires the `bookKeepingLock` and sets the state to "ShuttingDown". At that point `T2` might be calling execute and it might have checked the pool state before `T1` has set it to `ShuttingDown` but before it enqueues the task... 

So what then.. We will end up with one task in the queue that is kind of subjective whether it should be there or not. After all it was supplied to the pool before its state was transitioned to `ShuttingDown` or `ShutDown`. That is not a problem according to me. 

> What happens to enqueues which happen concurrently with shutdown?

They are rejected. If the pool has transitioned into any state different from `Running`, which is a matter of setting a `volatile` variable, the task will be rejected and that's it.